connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -430,7 +431,8 @@ public Void call() throws Exception {
                         if (!configState.contains(connName)) {
                             callback.onCompletion(new NotFoundException("Connector " + connName + " not found"), null);
                         } else {
-                            callback.onCompletion(null, new ConnectorInfo(connName, configState.connectorConfig(connName), configState.tasks(connName)));
+                            callback.onCompletion(null, new ConnectorInfo(connName, configState.connectorConfig(connName),
+                                configState.tasks(connName), worker.getConnectorType(connName)));
```

Comment:
Hmm.. Previously we only depended on the config state in order to get the connector info. This is available on all the workers even if they are not executing the connector. But now we need to get the type, which will only be available on the worker executing the connector. That seems less than ideal. It's kind of annoying, but maybe we have to load the class and check its type directly?

Reply:
Can you clarify the above suggestion ?
In current implementation, the worker knows the type of connector.

Reply:
The worker only maintains the state of the connectors that it is executing. A specific connector will only be running on one worker. The other workers will not have any state for the connector. So we will only be able to determine the connector type on the worker which is executing it.

Reply:
Looking at lines 204 and beyond in Worker#startConnector(), it seems having connector name is not enough to instantiate Connector.

I wonder how connector type request would be received by worker which doesn't have state for the connector.

Reply:
All nodes in the cluster have access to the global connector configuration and can respond to REST apis, but the `Worker` class only tracks the connectors being executed on that particular node. To find the type of a connector which is not executing on that node, we probably have to pull the classname out of the config using `ConnectorConfig.CONNECTOR_CLASS_CONFIG`.

Reply:
```
            final ConnectorConfig connConfig = new ConnectorConfig(plugins, connProps);
            final String connClass = connConfig.getString(ConnectorConfig.CONNECTOR_CLASS_CONFIG);
```
To get ConnectorConfig, connProps probably can be ignored.
What about plugins ?
Or, we can make connConfig a member of Worker so that we can get the connector class easily ?

Reply:
I am not sure. I would prefer to keep the current paradigm in which the worker only tracks the running connectors, but all the classloader logic makes it a little tricky to load the class from another context (I am not as familiar with this code). Maybe another option is to add the type to the configuration directly on creation since we already load the class in order to validate configuration and we already do some other config enrichment.

cc @ewencp In case you have any thoughts

Reply:
bq. add the type to the configuration directly

I was thinking about this as well.
You mean storing in ConnectorConfig object, right ?
However, such information is known to the Worker on that particular node.

Reply:
The Plugin system knows whether a connector is a source or a sink without having to create an instance. So, if we can get access to `Plugins`, we can get the connector plugins as a `Set<PluginDesc<Connector>>` set, find the plugin that matches the class name, and then look in the `PluginDesc.type()` to determine whether it's a source or sink.

So, if the current logic can't determine it because the connector's config state is not known locally, then the above could be a backup approach. Much of the time the state would be known.

Thoughts?

Reply:
Looks like PluginDesc only contains class information but not name.
While connectorInfo() is given connector name, not class.

...

Reply:
Yes, you'd need to find the name of the `Connector` implementation class for a given connector name. If we can't find that because we don't have the configuration, then we might just have to return null.